File: java/2034.股票价格波动.java
// 1. O(logn) t:117ms(64%) O(n) m:91.4M(74%) 哈希表+红黑树
class StockPrice {
    int maxTimestamp;
    HashMap<Integer, Integer> timePriceMap;
    TreeMap<Integer, Integer> prices;

    public StockPrice() {
        maxTimestamp = 0;
        timePriceMap = new HashMap<Integer, Integer>();
        prices = new TreeMap<Integer, Integer>();
    }
    
    public void update(int timestamp, int price) {
        maxTimestamp = Math.max(maxTimestamp, timestamp);
        if (timePriceMap.containsKey(timestamp)){
            int old = timePriceMap.get(timestamp);
            int cnt = prices.get(old);
            if (cnt == 1) prices.remove(old);
            else prices.put(old, cnt - 1);
        }
        timePriceMap.put(timestamp, price);
        prices.put(price, prices.getOrDefault(price, 0) + 1);
    }
    
    public int current() {
        return timePriceMap.get(maxTimestamp);
    }
    
    public int maximum() {
        return prices.lastKey();
    }
    
    public int minimum() {
        return prices.firstKey();
    }
}

// 2. O(logn) t:71ms(91%) O(n) m:91.9M(47%) 哈希表+双优先队列
class StockPrice {
    int maxTimestamp;
    HashMap<Integer, Integer> timePriceMap;
    PriorityQueue<int[]> pqMax;
    PriorityQueue<int[]> pqMin;

    public StockPrice() {
        maxTimestamp = 0;
        timePriceMap = new HashMap<Integer, Integer>();
        pqMax = new PriorityQueue<int[]>((a, b) -> b[0] - a[0]);
        pqMin = new PriorityQueue<int[]>((a, b) -> a[0] - b[0]);
    }
    
    public void update(int timestamp, int price) {
        maxTimestamp = Math.max(maxTimestamp, timestamp);
        timePriceMap.put(timestamp, price);
        pqMax.offer(new int[]{price, timestamp});
        pqMin.offer(new int[]{price, timestamp});
    }
    
    public int current() {
        return timePriceMap.get(maxTimestamp);
    }
    
    public int maximum() {
        while (true) {
            int[] priceTime = pqMax.peek();
            int price = priceTime[0], timestamp = priceTime[1];
            if (timePriceMap.get(timestamp) == price) {
                return price;
            }
            pqMax.poll();
        }
    }
    
    public int minimum() {
        while (true) {
            int[] priceTime = pqMin.peek();
            int price = priceTime[0], timestamp = priceTime[1];
            if (timePriceMap.get(timestamp) == price) {
                return price;
            }
            pqMin.poll();
        }
    }
}

File: python/2034.股票价格波动.py
# 1. O(logn) t:560ms(85%) O(n) m:59.7M(5%)
class StockPrice:

    def __init__(self):
        self.price = defaultdict(int)
        self.cur = -1
        self.min = []
        self.max = []

    def update(self, timestamp: int, price: int) -> None:
        if timestamp > self.cur: self.cur = timestamp
        self.price[timestamp] = price
        heapq.heappush(self.min, [price, timestamp])
        heapq.heappush(self.max, [-price, timestamp])

    def current(self) -> int:
        return self.price[self.cur]

    def maximum(self) -> int:
        while True:
            price, stamp = self.max[0]
            if -price == self.price[stamp]: return -price
            heapq.heappop(self.max)

    def minimum(self) -> int:
        while True:
            price, stamp = self.min[0]
            if price == self.price[stamp]: return price
            heapq.heappop(self.min)

from sortedcontainers import SortedList

# # [题解]2. O(logn) t:688ms(62%) O(n) m:52.3M(39%)
# class StockPrice:
#     def __init__(self):
#         self.price = SortedList()
#         self.timePriceMap = {}
#         self.maxTimestamp = 0

#     def update(self, timestamp: int, price: int) -> None:
#         if timestamp in self.timePriceMap:
#             self.price.discard(self.timePriceMap[timestamp])
#         self.price.add(price)
#         self.timePriceMap[timestamp] = price
#         self.maxTimestamp = max(self.maxTimestamp, timestamp)

#     def current(self) -> int:
#         return self.timePriceMap[self.maxTimestamp]

#     def maximum(self) -> int:
#         return self.price[-1]

#     def minimum(self) -> int:
#         return self.price[0]